fixed new room page form validation

ABee-Tech opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
Explained over here: https://github.com/Expensify/App/issues/6967#issuecomment-1007176416

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ [#6967](https://github.com/Expensify/App/issues/6967)

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Put empty room name
- [x] Existing room name check

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Open app
2. Create Workspace
3. Select "+" > New room
4. Enter any name and delete
5. Hit Create room with empty room name
6. User will not be able to create room with blank room name

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->


https://user-images.githubusercontent.com/42933600/148674612-57cf18a1-bbf8-4a6b-96bc-0386efe4f72d.mp4


https://user-images.githubusercontent.com/42933600/148674613-a6b9fed5-fb54-47d6-84ec-23066a6d9679.mp4

### Screenshots after all changes

#### Web

https://user-images.githubusercontent.com/42933600/154406834-3f78b44b-f41a-449b-b7a9-b7b23b5a7c22.mp4


#### Android


https://user-images.githubusercontent.com/42933600/154891211-7dd465b7-f963-41d5-aa08-88411d7e2d93.mp4

ABee-Tech committed:
fixed new room page form validation

ABee-Tech commented:
I think we should create new statements for: 
i. Please enter a room name.
i. Please select a workspace.
ii. Please select a visibility.

I need help on where to define these statements on `src/languages/en.js`.

ABee-Tech committed:
lint fix

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,107 @@ class WorkspaceNewRoomPage extends React.Component {
+        const modifiedRoomName = `#${modifiedRoomNameWithoutHash}`;
 
         const isExistingRoomName = _.some(
             _.values(this.props.reports),
-            report => report && report.policyID === this.state.policyID && report.reportName === finalRoomName,
+            report => report && report.policyID === this.state.policyID
+            && report.reportName === modifiedRoomName,
         );
```
Let's break this into a separate function which we can use in `validate`. This function will only be used to modify and store the room name.

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,107 @@ class WorkspaceNewRoomPage extends React.Component {
+            [inputKey]: value,
+            errors: {
+                ...prevState.errors,
+                [inputKey]: false,
```
```suggestion
                [inputKey]: '',
```

Keep the same datatype.

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -160,20 +217,21 @@ class WorkspaceNewRoomPage extends React.Component {
                             label={this.props.translate('newRoomPage.visibility')}
                             items={visibilityOptions}
-                            onChange={visibility => this.setState({visibility})}
+                            errorText={this.state.errors.visibility || ''}
```
```suggestion
                            errorText={this.state.errors.visibility}
```

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -160,20 +217,21 @@ class WorkspaceNewRoomPage extends React.Component {
                                 placeholder={{value: '', label: this.props.translate('newRoomPage.selectAWorkspace')}}
                                 items={this.state.workspaceOptions}
-                                onChange={this.onWorkspaceSelect}
+                                errorText={this.state.errors.policyID || ''}
```
```suggestion
                                errorText={this.state.errors.policyID}
```

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -160,20 +217,21 @@ class WorkspaceNewRoomPage extends React.Component {
                     </ScrollView>
                     <FixedFooter>
                         <Button
                             isLoading={this.props.isLoadingCreatePolicyRoom}
```
Please add `pressOnEnter`

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -149,9 +206,9 @@ class WorkspaceNewRoomPage extends React.Component {
+                            onChangeText={roomName => this.clearErrorAndSetValue('roomName', this.checkAndModifyRoomName(roomName).modifiedRoomName)}
                             value={this.state.roomName.substr(1)}
-                            errorText={this.state.error}
+                            errorText={this.state.errors.roomName || ''}
```
```suggestion
                            errorText={this.state.errors.roomName}
```

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -149,9 +206,9 @@ class WorkspaceNewRoomPage extends React.Component {
                             placeholder={this.props.translate('newRoomPage.social')}
                             containerStyles={[styles.mb5]}
-                            onChangeText={roomName => this.setState({roomName: this.checkAndModifyRoomName(roomName)})}
+                            onChangeText={roomName => this.clearErrorAndSetValue('roomName', this.checkAndModifyRoomName(roomName).modifiedRoomName)}
```
This will change, we should only be using the room name setter method
```suggestion
                            onChangeText={roomName => this.clearErrorAndSetValue('roomName', this.modifyRoomName(roomName))}
```

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,107 @@ class WorkspaceNewRoomPage extends React.Component {
+        }
+
+        if (!visibility) {
+            errors.visibility = 'Please select a visibility';
```
Use translations such as `this.props.translate('newRoomPage.roomAlreadyExists')`

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,107 @@ class WorkspaceNewRoomPage extends React.Component {
+        }
+
+        if (!policyID) {
+            errors.policyID = 'Please select a workspace';
```
use translations

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,107 @@ class WorkspaceNewRoomPage extends React.Component {
+        const errors = {};
+
+        if (!roomName || roomName === '#') {
+            errors.roomName = 'Please enter a room name';
```
use translations

ABee-Tech commented:
@parasharrajat can you tell me why is Jest Unit Tests is failing on something called `Decrypt OSBotify GPG Key`?

parasharrajat commented:
That is an internal issue and soon be fixed.

ABee-Tech commented:
And what about translation of the sentence to Español language? @parasharrajat

ABee-Tech committed:
refactor and translations

ABee-Tech commented:
For translation to Español language I have left empty string on es.js file.

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,104 @@ class WorkspaceNewRoomPage extends React.Component {
+
+        if (!roomName || roomName === '#') {
+            errors.roomName = this.props.translate('newRoomPage.pleaseEnterRoomName');
+        } else if (
+            policyID && isExistingRoomName
+        ) {
```
```suggestion
        } else if (policyID && isExistingRoomName) {
```

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,104 @@ class WorkspaceNewRoomPage extends React.Component {
+        const isExistingRoomName = _.some(
+            _.values(this.props.reports),
+            report => report && report.policyID === this.state.policyID
+            && report.reportName === this.modifyRoomName(roomName),
```
If we already modified the room name before storing it in the state then why do we need to modify it here again?

ABee-Tech replied:
Ohh yaa actually. I got confused haha.

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,104 @@ class WorkspaceNewRoomPage extends React.Component {
 
     /**
-     * Called when the "Create Room" button is pressed.
+     * Called when the onSubmit() method is triggered.
+     * which validates the form fields.
```
```suggestion
     * Validates the form fields and store errors in the state.
     * Returns false when validation fails.
```

parasharrajat replied:
OK. I see. I will revisit the style guide for a refresher. We can remove the comment then.

ABee-Tech replied:
Okay I will remove the comment then like this:
```
    /**
     * @returns {Boolean}
     */
```

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,104 @@ class WorkspaceNewRoomPage extends React.Component {
-     * Called when a workspace is selected. Also calls checkAndModifyRoomName,
-     * which displays an error if the given roomName exists on the newly selected workspace.
-     * @param {String} policyID
+     * Called when the "Create Room" button is pressed.
```
```suggestion
     * Create the room when the "Create Room" button is pressed.
```

ABee-Tech replied:
@marcaaron  I will remove entire comment for this method `onSubmit()` as per the jsdocs section on `STYLE.md`. And, what about naming-convention should I change the name from `onSubmit()` to `submit()` only?

parasharrajat replied:
`validateAndCreatePolicyRoom()` is suggested in next comment.

parasharrajat commented on src/languages/en.js:
```diff
@@ -825,6 +825,9 @@ export default {
         privateDescription: 'Only people invited to this room are able to find it',
         createRoom: 'Create Room',
         roomAlreadyExists: 'A room with this name already exists',
+        pleaseEnterRoomName: 'Please enter a room name',
+        pleaseSelectWorkspace: 'Please select a workspace',
+        pleaseSelectVisibility: 'Please select a visibility',
```
I am not too sure about these messages. 
I think the best would be to show. `{field} is a required field`.

So `${this.props.translate(fieldNameKey)} ${this.props.translate('common.isRequiredField')` is better for all.

cc: @marcaaron

marcaaron replied:
For copy we tag in @Expensify/marketing to give us the exact copy. Translations I think we can ask for help in Slack.

marcaaron replied:
The first two look good. I'm not sure about `Please select a visibility`, but think there should be no error since it's a required field and only has two options `Restricted | Private`.

jamesdeanexpensify replied:
Would the visibility be at the room-level? Maybe it could be something like "Please select a room type." Also, a couple questions:

1. Are these questions they'll get in this order? I'd imagine it would instead be more natural to have: workspace THEN room name THEN room type/visibility
2. Will we have descriptions of what "Restricted" and "Private" actually mean? Are there any other room types, or just these two?

marcaaron replied:
Thanks @jamesdeanexpensify! These are error messages that the user will see if they do not input correct values when they are creating a room.

To clarify, I meant to say "we don't need to worry about the visibility error" as there is a default selection made at all times (i.e. not possible to have a missing one).

jamesdeanexpensify replied:
Ah, understood - so "Restricted" and "Private" are defined for users elsewhere in the product? Just making sure. And for the conversation at hand, are you just looking for signoff on the proposed copy:

1. Please enter a room name
2. Please select a workspace

If so, then (personal preference) I'd probably update "select" to "choose" so it's "Please choose a workspace" but other than that it looks fine to me! 

Last thing - I'm going to ask around to see if we've been writing these error handling messages with periods at the end since I haven't been involved in creating many lately.

marcaaron replied:
>you just looking for signoff on the proposed copy

Yep! Thanks!

>I'm going to ask around to see if we've been writing these error handling messages with periods at the end since I haven't been involved in creating many lately.

Sounds good. I'm not sure either. Probably there is inconsistency and we can create an issue to resolve it.

jamesdeanexpensify replied:
Let's leave these without periods. Let me know if you need anything else, thanks!

marcaaron reviewed (changes requested):
Nice job so far! Left a few style reminders. Mostly looking good here.

ABee-Tech commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,104 @@ class WorkspaceNewRoomPage extends React.Component {
-    onWorkspaceSelect(policyID) {
-        this.setState({policyID});
-        this.checkAndModifyRoomName(this.state.roomName);
+    onSubmit() {
```
Great.

ABee-Tech commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,104 @@ class WorkspaceNewRoomPage extends React.Component {
+    validate() {
+        const roomName = this.state.roomName.trim();
+        const policyID = this.state.policyID;
+        const visibility = this.state.visibility;
```
Okay, Great.

ABee-Tech commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,104 @@ class WorkspaceNewRoomPage extends React.Component {
+            && report.reportName === this.modifyRoomName(roomName),
+        );
+
+        if (!roomName || roomName === '#') {
```
Ya, That's great.

ABee-Tech replied:
I will implement it.

ABee-Tech commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -77,50 +78,104 @@ class WorkspaceNewRoomPage extends React.Component {
+            errors.visibility = '';
+        }
+
+        if (!errors.roomName && !errors.policyID && !errors.visibility) {
```
> Instead of checking for each one of these, we can check for the `_.size(errors) > 0`

I will apply this method.

ABee-Tech replied:
> But I would recommend just settings `errors` to state and adding that check instead of returning `true` and `false`

Thats great. Thank you for the suggestion.

ABee-Tech committed:
added ROOM_PREFIX into CONST.js

ABee-Tech committed:
review changes

ABee-Tech commented:
@parasharrajat @marcaaron I have done changes as per reviews. You can review the changes.

parasharrajat commented on src/CONST.js:
```diff
@@ -537,6 +537,8 @@ const CONST = {
         EMOJIS: /(?:\uD83D(?:\uDC41\u200D\uD83D\uDDE8|\uDC68\u200D\uD83D[\uDC68\uDC69]\u200D\uD83D(?:\uDC66(?:\u200D\uD83D\uDC66)?|\uDC67(?:\u200D\uD83D[\uDC66\uDC67])?)|\uDC69\u200D\uD83D\uDC69\u200D\uD83D(?:\uDC66(?:\u200D\uD83D\uDC66)?|\uDC67(?:\u200D\uD83D[\uDC66\uDC67])?))|[\u2700-\u27bf]|(?:\ud83c[\udde6-\uddff]){2}|[\ud800-\udbff][\udc00-\udfff]|[\u0023-\u0039]\ufe0f?\u20e3|\u3299|\u3297|\ ...
     },
 
+    ROOM_PREFIX: '#',
```
Please move it under `POLICY`

parasharrajat commented on src/languages/en.js:
```diff
@@ -825,6 +825,9 @@ export default {
         createRoom: 'Create Room',
         roomAlreadyExists: 'A room with this name already exists',
+        pleaseEnterRoomName: 'Please enter a room name',
+        pleaseSelectWorkspace: 'Please select a workspace',
+        pleaseSelectVisibility: 'Please select a visibility',
```
Maybe replace `select` with `choose` and Ask for translations on slack.

parasharrajat commented on src/languages/es.js:
```diff
@@ -827,6 +827,9 @@ export default {
         privateDescription: 'Sólo las personas que están invitadas a esta sala pueden encontrarla',
         createRoom: 'Crea una sala de chat',
         roomAlreadyExists: 'Ya existe una sala con este nombre',
+        pleaseEnterRoomName: '',
+        pleaseSelectWorkspace: '',
+        pleaseSelectVisibility: '',
```
Need them...

marcaaron replied:
Can we ask in `#expensify-open-source` for correct translations? I'm not sure what the process is but someone should be able to help.

parasharrajat replied:
@ABee-Tech Asked them, waiting for reply...

ABee-Tech replied:
I have also asked on the channel.

parasharrajat reviewed (changes requested):
looks good. only the translation part is remaining.

ABee-Tech committed:
es translations and constant

ABee-Tech commented:
I have changed `select` to `choose` and finally added es translations for those error statements. I think everything is fine now.

marcaaron commented on src/languages/en.js:
```diff
@@ -825,6 +825,9 @@ export default {
         roomAlreadyExists: 'A room with this name already exists',
+        pleaseEnterRoomName: 'Please enter a room name',
+        pleaseChooseWorkspace: 'Please choose a workspace',
+        pleaseChooseVisibility: 'Please choose a visibility',
```
This can be removed. It is not possible to have no visibility set.

ABee-Tech replied:
Okay.

marcaaron commented on src/languages/es.js:
```diff
@@ -827,6 +827,9 @@ export default {
         roomAlreadyExists: 'Ya existe una sala con este nombre',
+        pleaseEnterRoomName: 'Por favor escribe el nombre de una sala',
+        pleaseChooseWorkspace: 'Por favor elige un espacio de trabajo',
+        pleaseChooseVisibility: 'Por favor elige una visibilidad',
```
Remove this line.

ABee-Tech replied:
Which line do you mean?

ABee-Tech replied:
ohh you mean `pleaseChooseVisibility`. Okay sure.

marcaaron replied:
Yes, sorry for not being specific. The comment usually refers to the line it was left on.

marcaaron commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -50,12 +50,13 @@ class WorkspaceNewRoomPage extends React.Component {
-        this.checkAndModifyRoomName = this.checkAndModifyRoomName.bind(this);
+        this.validate = this.validate.bind(this);
+        this.clearErrorAndSetValue = this.clearErrorAndSetValue.bind(this);
+        this.modifyRoomName = this.modifyRoomName.bind(this);
```
`this.modifyRoomName()` is the same. Please look and remove anything that doesn't need to be bound?

marcaaron replied:
`this.clearErrorAndSetValue` also seems to not need to be bound

marcaaron commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -76,51 +76,80 @@ class WorkspaceNewRoomPage extends React.Component {
+            errors.policyID = this.props.translate('newRoomPage.pleaseChooseWorkspace');
+        }
+
+        if (!this.state.visibility) {
+            errors.visibility = this.props.translate('newRoomPage.pleaseChooseVisibility');
+        }
```
```suggestion
```

ABee-Tech replied:
Okay.

marcaaron commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -76,51 +76,80 @@ class WorkspaceNewRoomPage extends React.Component {
+
+        if (!this.state.roomName || this.state.roomName === CONST.ROOM_PREFIX) {
+            errors.roomName = this.props.translate('newRoomPage.pleaseEnterRoomName');
+        } else if (this.state.policyID && isExistingRoomName) {
```
What is the significance of checking for `this.state.policyID`? Does it matter if we have one or not? If there is none then `isExistingRoomName` would always be `false`?

ABee-Tech replied:
I was doing that to confirm if workspace is first selected or not. But ya it's actually not needed. I will remove `this.state.policyID &&` from the condition.

marcaaron commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -76,51 +76,80 @@ class WorkspaceNewRoomPage extends React.Component {
+        }
+
+        this.setState({errors});
+        return Boolean(_.isEmpty(errors));
```
```suggestion
        return _.isEmpty(errors);
```

`_.isEmpty()` returns a Boolean so there is no need for the Boolean cast.

ABee-Tech replied:
Ok I will change it.

marcaaron commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -76,51 +76,80 @@ class WorkspaceNewRoomPage extends React.Component {
      * - Cannot not include space or special characters, and we automatically apply an underscore for spaces
      * - Must be lowercase
-     * Also checks to see if this room name already exists, and displays an error message if so.
+     * provides the modified room name.
```
```suggestion
```

marcaaron commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -160,22 +188,24 @@ class WorkspaceNewRoomPage extends React.Component {
                             label={this.props.translate('newRoomPage.visibility')}
                             items={visibilityOptions}
-                            onChange={visibility => this.setState({visibility})}
+                            errorText={this.state.errors.visibility}
```
```suggestion
```

This error should never happen as a default visibility is always set.

ABee-Tech replied:
Sure. Thanks.

marcaaron commented:
Found a few more things to fix. Also there are conflicts that will need to be fixed before we can merge.

ABee-Tech commented on src/CONST.js:
```diff
@@ -494,6 +494,8 @@ const CONST = {
         },
     },
 
+    ROOM_PREFIX: '#',
+
```
Oh I got it. You meant inside `POLICY`.

parasharrajat replied:
Yeah,

ABee-Tech commented on src/languages/es.js:
```diff
@@ -827,6 +827,9 @@ export default {
         createRoom: 'Crea una sala de chat',
         roomAlreadyExists: 'Ya existe una sala con este nombre',
+        pleaseEnterRoomName: 'Por favor escribe el nombre de una sala',
+        pleaseChooseWorkspace: 'Por favor elige un espacio de trabajo',
+        pleaseChooseVisibility: 'Por favor elige una visibilidad',
```
no they are for choose. You can see the [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1641913188292300). I asked for please choose.
![image](https://user-images.githubusercontent.com/42933600/149650977-57e24920-846e-4f16-b47e-66b86058e0af.png)

ABee-Tech replied:
@parasharrajat Are these translations incorrect?

ABee-Tech replied:
@parasharrajat @marcaaron what to do with this?

parasharrajat replied:
In the issue thread, marketing team suggested to use select instead of choose. So we should do that.

ABee-Tech committed:
review changes fixed

ABee-Tech commented:
@parasharrajat @marcaaron I have fixed all the review changes. Now, I think we should resolve conflicts first before merge. But, help me on resolving those conflicts.

parasharrajat commented:
@ABee-Tech Please merge Main into your branch and resolvee the conflicts if any. and let me know when this is ready.

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -76,51 +73,75 @@ class WorkspaceNewRoomPage extends React.Component {
         this.setState({workspaceOptions: _.map(workspaces, policy => ({label: policy.name, key: policy.id, value: policy.id}))});
     }
 
+    validateAndCreatePolicyRoom() {
```
NAB: This was supposed to be bound.

ABee-Tech replied:
Okay.

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -173,9 +194,9 @@ class WorkspaceNewRoomPage extends React.Component {
                             success
-                            onPress={this.onSubmit}
+                            pressOnEnter
+                            onPress={() => this.validateAndCreatePolicyRoom()}
```
NAB: This was correctly bound.

parasharrajat reviewed (changes requested):
Looks good. Please resolve the conflcits. thanks.

ABee-Tech committed:
validateAndCreatePolicyRoom binded

ABee-Tech commented:
@parasharrajat @marcaaron should we use `RoomNameInput` component used on line 203 of `WorkspaceNewRoomPage.js` or not?

ABee-Tech commented:
@parasharrajat @marcaaron Please help me rendering the room name error as there is no handling of `Please enter a room name`. And, errors are shown in `RoomNameInput` rather than on form component. Also, the behavior of showing error is set on `onChangeText` event rather than checking for errors after pressing `Create Room` button.

parasharrajat commented:
Where do you need help?
This is how I see it. 
1. Pass the error to RoomNameInput	via a prop. 
2. Move error handling for room name in the RoomNameInput.
3. Add another prop to determine if we want to show errors on change or on-demand. Default onChange, type Boolean.
4. Let RoomNameInput calculate the error onChange but don't show the error if point 3 prop say on-demand. 
5. onValidate, pass the error to RoomNameInput so that it can be shown.

marcaaron commented:
Can we get a status update here, please?

ABee-Tech commented:
Sorry for the delay. I will be resolving the conflicts by today.

ABee-Tech committed:
Merge branch 'main' of https://github.com/ABee-Tech/Expensify.cash into fix-empty-room-name

ABee-Tech commented:
@marcaaron @parasharrajat  there are some problems. Should I have to pass value from parent to child if I choose `showErrorOnDemand` to have full control of component. I am confused here. Because in this component value, onChange and errors are handled in child component i.e. `RoomNameInput`  I am pushing after some conflicts resolved. And, Please give me suggestion on changes. Should I change the behavior of other inputs or change only room name field with some extra tweaks like able to pass value of room name field from parent to child `RoomNameInput`? Please discuss on this.

ABee-Tech committed:
lint fix

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -26,6 +26,12 @@ const propTypes = {
     policyID: PropTypes.string,
 
+    /** Whether we should show the error on demand or not */
+    showErrorOnDemand: PropTypes.bool,
```
Something like `shouldShowErrorOnChange` is better.

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -106,15 +114,25 @@ class RoomNameInput extends Component {
-        // Certain names are reserved for default rooms and should not be used for policy rooms.
-        if (_.contains(CONST.REPORT.RESERVED_ROOM_NAMES, finalRoomName)) {
-            error = this.props.translate('newRoomPage.roomNameReservedError');
+        // We show the error provided as props if we want to show error on demand rather than on every change
+        if (this.props.errorText && this.props.showErrorOnDemand) {
+            error = this.props.errorText;
+        } else {
+            // We error if the user doesn't enter a room name or left blank
+            if (finalRoomName === CONST.POLICY.ROOM_PREFIX) {
+                error = this.props.translate('newRoomPage.pleaseEnterRoomName');
+            }
+
+            // We error if the room name already exists. We don't care if it matches the original name provided in this
+            // component because then we are not changing the room's name.
+            if (isExistingRoomName && finalRoomName !== this.originalRoomName) {
+                error = this.props.translate('newRoomPage.roomAlreadyExistsError');
+            }
+
+            // Certain names are reserved for default rooms and should not be used for policy rooms.
+            if (_.contains(CONST.REPORT.RESERVED_ROOM_NAMES, finalRoomName)) {
+                error = this.props.translate('newRoomPage.roomNameReservedError');
+            }
```
Here you just calculate the error as it was done before. You just need to add the new error cases from your changes.

Main change is in the JSX `errorText` prop of `TextInputWithPrefix`. 
it would become something like that.
```
errorText={shouldShowErrorOnchange ? this.state.error : this.props.errorText}
```

ABee-Tech replied:
nice

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -115,18 +172,22 @@ class WorkspaceNewRoomPage extends React.Component {
                         <View style={styles.mb5}>
                             <Text style={[styles.formLabel]}>{this.props.translate('newRoomPage.roomName')}</Text>
                             <RoomNameInput
-                                onChangeText={(roomName) => { this.setState({roomName}); }}
                                 initialValue={this.state.roomName}
                                 policyID={this.state.policyID}
+                                showErrorOnDemand
+                                errorText={this.state.errors.roomName}
+                                onChangeText={roomName => this.clearErrorAndSetValue('roomName', this.modifyRoomName(roomName))}
```
Here, you take the error calculated on `RoomNameInput` and save it on state via `onChangeError`. 
And on submit, you will pick the error and move it to `this.state.errors.roomName` instead of calculating the same here. 
Rest would be the same as before. 

We do all this to keep the error logic on the `RoomNameInput` component and still show the error only onSubmit. But `RoomNameInput` is used somewhere else which shows the error on change of value.

parasharrajat replied:
There could be other ways. Please thing about it.

ABee-Tech replied:
In this way we don't have access of blank room name error at first before changing the `roomName` state as the error validate will run on `RoomNameInput` only if there is change in room name state. I think this much is good by keeping  logic on both `RoomNameInput` and `WorkspaceNewRoomPage`.

ABee-Tech replied:
Why is there red outline on field when error is shown and not in other fields? There is no handling for that on `Picker`.

ABee-Tech replied:
@parasharrajat what can I do for it?

parasharrajat replied:
there is `hasError` prop on the Picker which sets the outline.

ABee-Tech committed:
showErrorOnDemand on RoomNameInput

parasharrajat commented:
@ABee-Tech Could you please ping me when this is ready?

ABee-Tech committed:
hasError props on picker for outline color

ABee-Tech commented:
@parasharrajat Everything is ready to deploy.

parasharrajat commented:
I looked at the other place where RoomNameInput is used. 
![image](https://user-images.githubusercontent.com/24370807/151148449-59c68beb-99e9-4d46-a10c-c039bce3deea.png) 

I think we need to use the normal on submit form error flow for this as well. I am not sure about this but that is certainly out of the scope of this issue. Let's wait for Marc to reply on this.

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -76,19 +76,85 @@ class WorkspaceNewRoomPage extends React.Component {
+        }));
+    }
+
+    /**
+     * Modifies the room name to follow our conventions:
+     * - Max length 80 characters
+     * - Cannot not include space or special characters, and we automatically apply an underscore for spaces
+     * - Must be lowercase
+     * @param {String} roomName
+     *
+     * @returns {String}
+     */
+    modifyRoomName(roomName) {
+        const modifiedRoomNameWithoutHash = roomName.substr(1)
+            .replace(/ /g, '_')
+            .replace(/[^a-zA-Z\d_]/g, '')
+            .substr(0, CONST.REPORT.MAX_ROOM_NAME_LENGTH)
+            .toLowerCase();
+        const modifiedRoomName = `${CONST.POLICY.ROOM_PREFIX}${modifiedRoomNameWithoutHash}`;
+
+        return modifiedRoomName;
```
The same thing exists on RoomNameInput. Could you please dry this up?

ABee-Tech replied:
Sure

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -76,19 +76,85 @@ class WorkspaceNewRoomPage extends React.Component {
      */
-    onWorkspaceSelect(policyID) {
-        this.setState({policyID});
+    validate() {
+        const errors = {};
+
+        const isExistingRoomName = _.some(
+            _.values(this.props.reports),
+            report => report && report.policyID === this.state.policyID
+            && report.reportName === this.state.roomName,
+        );
+
+        // We error if the user doesn't enter a room name or left blank
+        if (!this.state.roomName || this.state.roomName === CONST.POLICY.ROOM_PREFIX) {
+            errors.roomName = this.props.translate('newRoomPage.pleaseEnterRoomName');
+        }
+
+        // We error if the room name already exists. We don't care if it matches the original name provided in this
+        // component because then we are not changing the room's name.
+        if (isExistingRoomName) {
+            errors.roomName = this.props.translate('newRoomPage.roomAlreadyExistsError');
+        }
+
+        // Certain names are reserved for default rooms and should not be used for policy rooms.
+        if (_.contains(CONST.REPORT.RESERVED_ROOM_NAMES, this.state.roomName)) {
+            errors.roomName = this.props.translate('newRoomPage.roomNameReservedError');
+        }
+
+        if (!this.state.policyID) {
+            errors.policyID = this.props.translate('newRoomPage.pleaseSelectWorkspace');
+        }
+
+        this.setState({errors});
+        return _.isEmpty(errors);
```
Please dry this as well.

ABee-Tech replied:
We can dry this up but code will become messy. I have defined why we cannot dry validate part [here](https://github.com/Expensify/App/pull/7098#discussion_r790137091).

ABee-Tech replied:
@parasharrajat please look at the comment and give me reply or suggestions on it.

ABee-Tech replied:
@parasharrajat what should I do for that one?

parasharrajat reviewed (changes requested):
Overall changes are fine. But we should think to dry this up. Reduce duplicate code.

ABee-Tech committed:
dried up modifyRoomName part

ABee-Tech committed:
Merge branch 'main' of https://github.com/ABee-Tech/Expensify.cash into fix-empty-room-name

ABee-Tech committed:
lint fix

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -82,45 +90,52 @@ class RoomNameInput extends Component {
 
         let error = '';
 
+
+        // We error if the user doesn't enter a room name or left blank
+        if (modifiedRoomName === CONST.POLICY.ROOM_PREFIX) {
+            error = this.props.translate('newRoomPage.pleaseEnterRoomName');
+        }
+
```
I think we don't need this error here. onchange validation are added to RoomNameInput for Room Settings page where we expect to show the error as user types. On that page, save button is disabled when there is no input to roominput. Thus this error does not make sense there. 

Also,  as we are separately checking on WorkspaceNewRoompage.

parasharrajat replied:
```suggestion
```

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -82,45 +90,52 @@ class RoomNameInput extends Component {
+     * - Cannot not include space or special characters, and we automatically apply an underscore for spaces
+     * - Must be lowercase
+     * @param {String} roomName
+     *
```
```suggestion
```

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -82,45 +90,52 @@ class RoomNameInput extends Component {
-            .toLowerCase();
-        const finalRoomName = `#${modifiedRoomNameWithoutHash}`;
-
+    validateRoomName(modifiedRoomName) {
```
Missing Jsdoc

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -69,7 +77,7 @@ class RoomNameInput extends Component {
         this.originalRoomName = props.initialValue;
 
-        this.checkAndModifyRoomName = this.checkAndModifyRoomName.bind(this);
+        this.validateRoomName = this.validateRoomName.bind(this);
```
Not needed. 
```suggestion

```

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -26,6 +26,12 @@ const propTypes = {
     policyID: PropTypes.string,
 
+    /** Whether we should show the error on demand or not */
+    shouldShowErrorOnDemand: PropTypes.bool,
```
```suggestion
    shouldShowErrorOnChange: PropTypes.bool,
```

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -26,6 +26,12 @@ const propTypes = {
     /** ID of policy whose room names we should be checking for duplicates */
     policyID: PropTypes.string,
 
+    /** Whether we should show the error on demand or not */
```
```suggestion
    /** Whether to show the error on change of input or not */
```

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -56,6 +62,8 @@ const defaultProps = {
     initialValue: '',
     disabled: false,
     policyID: '',
+    shouldShowErrorOnDemand: false,
```
```suggestion
    shouldShowErrorOnChange: true,
```

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -131,9 +146,13 @@ class RoomNameInput extends Component {
+                }}
                 value={this.state.roomName.substring(1)}
-                errorText={this.state.error}
+                errorText={this.props.shouldShowErrorOnDemand ? this.props.errorText : this.state.error}
```
```suggestion
                errorText={this.props.shouldShowErrorOnChange ? this.state.error: this.props.errorText}
```

parasharrajat commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -115,18 +160,23 @@ class WorkspaceNewRoomPage extends React.Component {
-                                onChangeText={(roomName) => { this.setState({roomName}); }}
                                 initialValue={this.state.roomName}
                                 policyID={this.state.policyID}
+                                shouldShowErrorOnDemand
```
```suggestion
                                shouldShowErrorOnChange={false}
```

parasharrajat reviewed (changes requested):
Looks good. Just some general tweaks. Let me know when ready. Marc can help us dry the validations. 
A suggestion could be to export this validation function to lib and then run the specific validations that are needed by passing parameters.

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -131,9 +146,13 @@ class RoomNameInput extends Component {
+                onChangeText={(roomName) => {
+                    const modifiedRoomName = this.modifyRoomName(roomName);
+                    this.setState({roomName: modifiedRoomName});
+                    this.validateRoomName(modifiedRoomName);
```
Also, we can skip this function when `this.props.shouldShowErrorOnDemand` is false. 
```suggestion
		 if(!this.props.shouldShowErrorOnDemand){
		 	return;	
		 }
                 
                 this.validateRoomName(modifiedRoomName);
```

ABee-Tech committed:
review fixes

ABee-Tech commented:
@parasharrajat Done applying changes.

ABee-Tech committed:
lint fix

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -26,10 +26,10 @@ const propTypes = {
+    shouldShowErrorOnChange: PropTypes.bool,
 
-    /** On demand error text if set shouldShowErrorOnDemand true */
+    /** On demand error text if set shouldShowErrorOnChange true */
```
```suggestion
    /** Error text to show when  shouldShowErrorOnChange is false. */
```

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -83,44 +89,47 @@ class RoomNameInput extends Component {
+            .replace(/[^a-zA-Z\d_]/g, '')
+            .substr(0, CONST.REPORT.MAX_ROOM_NAME_LENGTH)
+            .toLowerCase();
+        const modifiedRoomName = `${CONST.POLICY.ROOM_PREFIX}${modifiedRoomNameWithoutHash}`;
+
+        return modifiedRoomName;
```
```suggestion
        return `${CONST.POLICY.ROOM_PREFIX}${modifiedRoomNameWithoutHash}`;
```

parasharrajat reviewed (changes requested):
Close to finish.

ABee-Tech committed:
some fixes

parasharrajat reviewed (dismissed):
LGTM. 

cc: @marcaaron 

:ribbon: :eyes: :ribbon:  C+ reviewed

parasharrajat commented:
@ABee-Tech Please test it on the rest of the platforms, tick the checkboxes and attach screenshots.

ABee-Tech commented:
@parasharrajat Done.

ABee-Tech commented:
I don't have iOS device and Mac Device. So, I need help testing them in iOS and Desktop App. 
And, @parasharrajat How can we test for web? which command to run?

parasharrajat commented:
This information is available in README.md file. In short run `npm run web`.

ABee-Tech commented:
Yes, I have tried `npm run web`. But, it always shows this error:
![image](https://user-images.githubusercontent.com/42933600/151968340-caf9c05d-415e-47da-b8d3-023ee27fe87b.png)

Do you know how to solve it?

parasharrajat commented:
Best place to discuss this is the Slack open source channel. Thanks.

marcaaron commented on src/CONST.js:
```diff
@@ -519,8 +519,10 @@ const CONST = {
+        ROOM_PREFIX: '#',
     },
 
+
```
Please remove the extra new line here.

marcaaron commented on src/components/RoomNameInput.js:
```diff
@@ -26,6 +26,12 @@ const propTypes = {
     policyID: PropTypes.string,
 
+    /** Whether to show the error on change of input or not */
+    shouldShowErrorOnChange: PropTypes.bool,
```
Not sure I understand where this is coming from. Why would we _not_ want to validate on change? We should aim to keep this simple because our form component currently in development will soon replace logic like this...

parasharrajat replied:
I asked the question about this https://github.com/Expensify/App/pull/7098#issuecomment-102207487 to you.

Basically, Room SettingsPage shows the error on the change of the input. So to preserve that behavior and still utilize this component, this is the easiest way. 

If you say that we should also make that like other forms of the app, I would say that would be out of the scope of this issue.

marcaaron replied:
Not seeing what comment you're referring to - I think GH is having issues atm.

>I would say that would be out of the scope of this issue.

If we are adding things that don't need to be added then it's in scope 😄 

>Room SettingsPage shows the error on the change of the input

Where does this happen? Am I missing it? Can you give me an example screenshot or video? 

WRT form handling no errors should be shown on change unless the input is blurred first. I'm not sure it is worth preserving a behavior that will be changed very soon. Also, the entire feature is behind a beta?

marcaaron replied:
Regardless, it should be up to the parent I think. This component should be dumb and just show whatever error is passed to it.

parasharrajat replied:
Yeah, validation should be handled via Parent but this component was designed to be reactive for https://github.com/Expensify/App/blob/9d6e07a0d3bb298d7edba16824efb95afc94f280/src/pages/ReportSettingsPage.js#L133-L154. 

To keep it compatible with that page I suggested using this prop.

But I showed my concern that the Settings page input also show error on submit like other forms.
![image](https://user-images.githubusercontent.com/24370807/152276871-7c1c2d7b-173b-4fff-984a-7e4e6360a0cb.png)

marcaaron replied:
What does "designed to be reactive" mean? Pretty sure errors should be passed to inputs and not have `onChangeError` callbacks. Seems like we maybe got the design wrong on this component.

@ABee-Tech can you please remove this prop for now?

ABee-Tech replied:
Okay, I also think onChangeError has no work as we are rendering error inside the component. I will remove `onChangeError`.

ABee-Tech replied:
@marcaaron The `onChangeError` is used for disabling Save button to check whether there is error or not on `ReportSettingsPage`. 

https://github.com/Expensify/App/blob/1d72a5e01ed6e0c0b500f7e0617d05eb8c05dee9/src/pages/ReportSettingsPage.js#L132-L155

So, I think we should change behavior of `ReportSettingsPage` also by changing validation from on change text to on form submit. This will cancel button disabling and change behavior as it should be.

marcaaron commented on src/components/RoomNameInput.js:
```diff
@@ -26,6 +26,12 @@ const propTypes = {
+    /** Whether to show the error on change of input or not */
+    shouldShowErrorOnChange: PropTypes.bool,
+
+    /** Error text to show when shouldShowErrorOnChange is false */
```
It's just error text. We shouldn't refer to any of the component logic in our description of this prop.

```suggestion
    /** Error text to show */
```

marcaaron commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -115,18 +160,23 @@ class WorkspaceNewRoomPage extends React.Component {
+                                onChangeText={roomName => this.clearErrorAndSetValue('roomName', roomName)}
                             />
                         </View>
+
```
remove newline

marcaaron reviewed (changes requested):
Looks good, but found a few small issues also not sure why we need `shouldShowErrorOnChange`.

ABee-Tech commented:
Testing on Web:

https://user-images.githubusercontent.com/42933600/152171225-2f50294e-f836-403e-a3c2-b45d2cd12fcc.mp4

ABee-Tech committed:
comments and spacing fixes

marcaaron commented on src/components/RoomNameInput.js:
```diff
@@ -83,44 +89,46 @@ class RoomNameInput extends Component {
-            .toLowerCase();
-        const finalRoomName = `#${modifiedRoomNameWithoutHash}`;
-
+    validateRoomName(modifiedRoomName) {
```
Validation should happen in whatever parent component is using this. If we need to re-use it we should turn it into a utility method not have validation happen inside the input. Looking at this component again and the design is not compatible with our new forms and must change.

ABee-Tech replied:
@marcaaron Sure I will work on it. I have mentioned the same thing on [this comment](https://github.com/Expensify/App/pull/7098#issuecomment-1018188706) that validation is on child input component. That should be transferred to parent form component. So, what should we do Marc? Should we use validation on parent component only on where ever the `RoomNameInput` component are implemented? or make utility function on `libs` folder?

parasharrajat replied:
There is the same set of validations for room names everywhere. So you can move these to a utility and then if you need to skip a few validations then that can be achieved vai params.

marcaaron replied:
Yes 

- remove the validation from the input and move it to `ValidationUtils.js`
- use this to validate in the parent component for each case
- See various methods used to validate input for reference...

https://github.com/Expensify/App/blob/ed588095ee477c489b5278f1bd251476b5281261/src/libs/ValidationUtils.js#L324-L346

marcaaron replied:
>Should we use validation on parent component only on where ever the RoomNameInput component are implemented

Yes

parasharrajat replied:
Could you please do your changes and then we can review those?

marcaaron replied:
@ABee-Tech I'm not sure that name makes sense as there is another error related to the `policyID` and that validation should be separate from something called `isValidRoomName()`.

ABee-Tech replied:
Okay that makes sense. Thanks. I will apply the changes and will inform.

ABee-Tech replied:
I have done changes. Please review them.

marcaaron commented on src/components/RoomNameInput.js:
```diff
@@ -131,9 +139,16 @@ class RoomNameInput extends Component {
+                    if (!this.props.shouldShowErrorOnChange) {
+                        return;
+                    }
+                    this.validateRoomName(modifiedRoomName);
```
None of this should be happening in an "input" component.

marcaaron commented on src/components/RoomNameInput.js:
```diff
@@ -131,9 +139,16 @@ class RoomNameInput extends Component {
+                }}
                 value={this.state.roomName.substring(1)}
-                errorText={this.state.error}
+                errorText={this.props.shouldShowErrorOnChange ? this.state.error : this.props.errorText}
```
Error text should come from the parent not be set to state. Inputs should not handle their own validation.

marcaaron reviewed (changes requested):
I think we should go ahead and fix the other usage of `RoomNameInput` so that it is consistent with our other form inputs.

@ABee-Tech let me know if you understand the refactor I'm asking for and if not I can explain it in detail. We can double the fee if you are interested and can do the refactor before we merge this PR. Please let me know if that's possible.

ABee-Tech committed:
isReservedRoomName and isExistingRoomName added to ValidationUtils

ABee-Tech committed:
change form behaviour and apply validator on parent

ABee-Tech commented:
I have completed the suggested changes. Please review them.
### Changes:
1. I have added `isReservedRoomName` and `isExistingRoomName` utility functions on `ValidationUtils.js`. 
2. Applied those validations into components where `RoomNameInput` is placed. 
3. Removed validation from `RoomNameInput`, moved to whichever parent component the component is placed and changed form behavior.

ABee-Tech committed:
Merge branch 'main' of https://github.com/ABee-Tech/Expensify.cash into fix-empty-room-name

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -72,45 +59,17 @@ class RoomNameInput extends Component {
-
-        this.setState({error});
+        this.props.onChangeText(this.state.roomName);
     }
```
We don't need this block now. Call the `this.props.onChangeText(this.state.roomName);` directly when value changes.

ABee-Tech replied:
We need the state for room name modification.

parasharrajat commented on src/components/RoomNameInput.js:
```diff
@@ -142,13 +101,9 @@ class RoomNameInput extends Component {
-                    if (!this.props.shouldShowErrorOnChange) {
-                        return;
-                    }
-                    this.validateRoomName(modifiedRoomName);
```
Here. Call `this.props.onChangeText(this.state.roomName);` and I also think we don't need the state for this component as well.

`this.setState({roomName: modifiedRoomName});` This seems unnecessary now.

ABee-Tech replied:
We need the state for modified room name.

parasharrajat replied:
Ok. let's just focus on `componentDidupdate` part for now. we don't need it. Call `this.props.onChangeText(this.state.roomName);` here.

ABee-Tech commented on src/pages/ReportSettingsPage.js:
```diff
@@ -84,12 +92,58 @@ class ReportSettingsPage extends Component {
+    validate() {
+        const errors = {};
+
+        // We error if the user doesn't enter a room name or left blank
+        if (!this.state.newRoomName || this.state.newRoomName === CONST.POLICY.ROOM_PREFIX) {
+            errors.newRoomName = this.props.translate('newRoomPage.pleaseEnterRoomName');
+        }
+
+        // We error if the room name already exists. We don't care if it matches the original name provided in this
+        // component because then we are not changing the room's name.
+        if (ValidationUtils.isExistingRoomName(this.state.newRoomName, this.props.reports, this.props.report.policyID)) {
+            errors.newRoomName = this.props.translate('newRoomPage.roomAlreadyExistsError');
+        }
+
+        // Certain names are reserved for default rooms and should not be used for policy rooms.
+        if (ValidationUtils.isReservedRoomName(this.state.newRoomName)) {
+            errors.newRoomName = this.props.translate('newRoomPage.roomNameReservedError');
+        }
+
+        this.setState({errors});
+        return _.isEmpty(errors);
+    }
+
+    /**
+     * @param {String} inputKey
+     * @param {String} value
+     */
+    clearErrorAndSetValue(inputKey, value) {
+        this.setState(prevState => ({
+            [inputKey]: value,
+            errors: {
+                ...prevState.errors,
+                [inputKey]: '',
+            },
+        }));
     }
 
     render() {
-        const shouldDisableRename = ReportUtils.isDefaultRoom(this.props.report) || ReportUtils.isArchivedRoom(this.props.report) || this.props.isLoadingRenamePolicyRoom;
+        const shouldDisableRename = ReportUtils.isDefaultRoom(this.props.report) || ReportUtils.isArchivedRoom(this.props.report);
```
The reason behind this change is we dont have to disable button while loading rename policy room because it is disabled inside Button component with the help of isLoading prop.

ABee-Tech commented on src/pages/ReportSettingsPage.js:
```diff
@@ -131,26 +185,23 @@ class ReportSettingsPage extends Component {
                             <Button
-                                success={!shouldDisableRename}
+                                success
+                                pressOnEnter
```
Which idea you are talking about?

parasharrajat replied:
`pressOnEnter`

ABee-Tech commented on src/pages/ReportSettingsPage.js:
```diff
@@ -84,12 +92,58 @@ class ReportSettingsPage extends Component {
+        return _.isEmpty(errors);
+    }
+
+    /**
+     * @param {String} inputKey
+     * @param {String} value
+     */
+    clearErrorAndSetValue(inputKey, value) {
+        this.setState(prevState => ({
+            [inputKey]: value,
+            errors: {
+                ...prevState.errors,
+                [inputKey]: '',
+            },
+        }));
     }
```
I want to keep it because it will bring consistency to the forms.

parasharrajat commented on src/pages/ReportSettingsPage.js:
```diff
@@ -49,6 +50,13 @@ const propTypes = {
 
+    /** All reports shared with the user */
+    reports: PropTypes.shape({
+        reportName: PropTypes.string,
+        type: PropTypes.string,
+        policyID: PropTypes.string,
```
Please add the comments. You can read more about this on the STYLE.md

ABee-Tech committed:
suggested fixes

ABee-Tech committed:
lint fix

parasharrajat reviewed (changes requested):
Please test it on the remaining platforms and add screenshots.

parasharrajat reviewed (dismissed):
Changes look good to me. Only screenshots are remaining. 

cc: @marcaaron 

:ribbon: :eyes: :ribbon:  C+ reviewed

marcaaron commented:
>Only screenshots are remaining.

@parasharrajat can you ping me again when this is 100% ready please, thanks!

ABee-Tech commented:
I need help on ios and mac. I will be importing android and web by today.

ABee-Tech commented:
@parasharrajat  I need help on ios and mac. I will be importing android and web by today.

ABee-Tech commented:
@parasharrajat One last fix I forgot to fix. That is not showing `The room name already exists` when trying to rename policy with same name as its existing name from `ReportSettingsPage`. I will fix it and upload screenshots.

ABee-Tech commented:
@parasharrajat But the error from server also says A room with this name already exists. so should I keep it on validation level or leave to server to show error.

parasharrajat commented:
The slack post didn't get much attention but it seems we can proceed with the [changes](https://github.com/Expensify/App/pull/7098#issuecomment-1035070164).

ABee-Tech commented:
Okay then.

ABee-Tech committed:
no action on same name as previous

ABee-Tech committed:
Merge branch 'main' of https://github.com/ABee-Tech/Expensify.cash into fix-empty-room-name

ABee-Tech committed:
lint fix

parasharrajat reviewed (changes requested):
Thanks for the changes. I noticed some changed code in your PR. Please make sure that you don't break the fix https://github.com/Expensify/App/pull/7661. 

Please adjust the code so that 7661 still works on your changes. And It would be great if you can mark old comments resolved. Thanks.


Modify your screenshots to reflect both changes. new Room page and Room Settings page in the single video for each platform. You can leave Desktop and IOS, I will test that as you don't have access to it.

ABee-Tech commented:
Everything is fine. There is no need for change I think. That works fine with our code. I will attach screenshots for Android and Web.

ABee-Tech commented:
@parasharrajat Everything is fine. There is no need for change I think. That works fine with our code. I will attach screenshots for Android and Web.

ABee-Tech committed:
event handled for cursor movement fix

ABee-Tech commented:
@parasharrajat I have adjusted the code and tested on web and android. I have uploaded screenshots above.

parasharrajat commented on src/pages/ReportSettingsPage.js:
```diff
@@ -84,12 +97,58 @@ class ReportSettingsPage extends Component {
+    }
+
+    validateAndRenameReport() {
+        if (!this.validate() || this.props.report.reportName === this.state.newRoomName) {
```
It seems that showing a successful growl instead of skipping when the user uses the same name is better UX. Could you please update that? It is also mentioned on the slack.

parasharrajat replied:
Without making an API call, just show the success growl.

parasharrajat reviewed (changes requested):
Thanks, @ABee-Tech. One more small tweak. 

![image](https://user-images.githubusercontent.com/24370807/154445589-36828da3-5bea-43f4-ab09-c08d37400720.png)
Please tick all platforms.

ABee-Tech committed:
suggested changes

ABee-Tech committed:
lint fix and added comments

parasharrajat commented on src/pages/ReportSettingsPage.js:
```diff
@@ -84,12 +98,62 @@ class ReportSettingsPage extends Component {
+    }
+
+    validateAndRenameReport() {
+        if (this.props.report.reportName === this.state.newRoomName) {
+            Growl.success(this.props.translate('newRoomPage.policyRoomRenamed'));
+            return;
+        }
+        if (!this.validate()) {
+            return;
+        }
```
```suggestion
        if (this.props.report.reportName === this.state.newRoomName) {
            Growl.success(this.props.translate('newRoomPage.policyRoomRenamed'));
            return;
        }
        if (!this.validate()) {
            return;
        }
```

Please move the validate block above the growl one.

parasharrajat reviewed (changes requested):
One change and we should be good.

ABee-Tech commented:
@parasharrajat I am having trouble running web version. Can we consider last screenshot just without growl message?

ABee-Tech committed:
moved validate block above growl

parasharrajat reviewed (dismissed):
LGTM. Thanks for the changes @ABee-Tech 

cc: @marcaaron 

:ribbon: :eyes: :ribbon:  C+ reviewed

ABee-Tech commented:
Android screenshot updated.

marcaaron commented on src/libs/ValidationUtils.js:
```diff
@@ -321,6 +321,34 @@ function doesFailCharacterLimit(maxLength, valuesToBeValidated) {
+ * component because then we are not changing the room's name
+ *
+ * @param {String} roomName
+ * @param {Object[]} reports
```
```suggestion
 * @param {Object} reports
```

marcaaron commented on src/libs/ValidationUtils.js:
```diff
@@ -321,6 +321,34 @@ function doesFailCharacterLimit(maxLength, valuesToBeValidated) {
+ */
+function isExistingRoomName(roomName, reports, policyID) {
+    return _.some(
+        _.values(reports),
```
`_.some()` should work on object collections so no need to use `_.values()` here 

https://underscorejs.org/#some

marcaaron commented on src/pages/ReportSettingsPage.js:
```diff
@@ -128,26 +192,22 @@ class ReportSettingsPage extends Component {
-                                    || this.state.error,
-                                )}
                                 isLoading={this.props.isLoadingRenamePolicyRoom}
+                                isDisabled={shouldDisableRename}
```
NAB (as we are conducting an audit of all forms now) but no form buttons should be disabled

parasharrajat replied:
I think this needs to be disabled for this case when user is seeing settings for default policies #admins, etc. I haven't seen the new Form style draft so I can't think of anything apart from it.

marcaaron replied:
We'll release the guidelines soon. Basically, no forms should be disabled for any reason.

parasharrajat replied:
Ok. Thanks.

marcaaron commented on src/pages/workspace/WorkspaceNewRoomPage.js:
```diff
@@ -76,19 +82,59 @@ class WorkspaceNewRoomPage extends React.Component {
+        }
+
+        // We error if the room name already exists. We don't care if it matches the original name provided in this
+        // component because then we are not changing the room's name.
```
I don't really understand the comment. In particular...

>We don't care if it matches the original name provided in this component because then we are not changing the room's name.

What does this have to do with the check we make? How is it determined that the name in state matches the "original name provided in this component"? Which variable represents that "original name"? Can we be more specific or is the comment out of context?

ABee-Tech replied:
@marcaaron Yes actually it is out of context for this page so we should remove this and we can be more specific on same comment on `ReportSettingsPage` by just modifing it as `We don't error if the room name matches same as previous.`

ABee-Tech committed:
changes in comments and minor fixes

ABee-Tech commented:
@marcaaron Please review the changes.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1889649076) to staging by @marcaaron in version: 1.1.40-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1917947501) to production by @yuwenmemon in version: 1.1.40-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

Stutikuls commented:
**Issue 1 -**

**Title**- [Medium] Chrome +Jaws : 'Room Name' text edit field is not marked as required and announced.
**Actual** - The fields 'Room name' text field is required to be filled but is not announced as required and no visual "*" sign for required field.
Screen reader only read the value of the Room name, not reading the Label 'Room name' along with the value.
**Expected** -If the form element is required, it should be announced as such for the screen reader users and visual "*" sign for required field should be there. Screen reader should read like "Abc  room name editable text box".
**WCAG failure** - 1.3.1
**Reproducible in staging?**: - Yes
**Version Number**: - v1.1.40-2
**Platforms** - Web (Chrome + Jaws), Desktop (MAC), iOS(Voiceover), Mobile - web(Safari + Voiceover)

**Note** - Due to installation of build not able to check on Android


https://user-images.githubusercontent.com/97600950/157184775-555e29e3-b3b8-484d-975b-253176be91a4.mp4


**Issue 2 -** 

**Title**-[Medium] Chrome + Jaws: Screen reader : Error message is not announced by the screen reader.
**Actual** - Screen reader does not announce any inline error message appear below to text boxes.
**Expected** - Inline error message should be conveyed to the users as soon as it appear on the fields so that user can resolve the error.
**WCAG failure** - 3.3.1
**Reproducible in staging?**: - Yes
**Version Number**: - v1.1.40-2
**Platforms** - Web(Chrome + Jaws), Desktop(MAC), Mobile-web (Safari + Voiceover), iOS(Voiceover)

**Note** - Due to installation of build not able to check on Android

https://user-images.githubusercontent.com/97600950/157184865-5b1d8ce7-aa37-494f-b14a-b993685a6fdd.mp4



**Issue 3 -** 

**Title**-[Medium] Chrome + Jaws: Screen reader : Screen reader is not provide any information after creating the room.
**Actual** - Screen reader does not announce any information after creating the room, screen reader is reading some extra hidden information.
**Expected** - After creating the room, focus should lands on the name of the room(new) and should read the new created room name.
**WCAG failure** - 1.3.1
**Reproducible in staging?**: - Yes
**Version Number**: - v1.1.40-2
**Platforms** - Web(Chrome + Jaws), Desktop(MAC), Mobile-web (Safari + Voiceover), iOS(Voice over)

**Note** - Due to installation of build not able to check on Android

https://user-images.githubusercontent.com/97600950/157185999-74c62343-2664-4b73-8c76-635ee832cd72.mp4


**Issue 4 -** 

**Title**-[Medium] Safari+ Voiceover : Screen reader : Screen reader is not reading role for 'create room' control.
**Actual** - Screen reader only announce "create room".
              **Note -** On Mac reading "create room Group" when focus lands on the create room control.
**Expected** - Role should define as 'Button' and screen reader should read like "Create room button. 
**WCAG failure** - 4.1.2
**Reproducible in staging?**: - Yes
**Version Number**: - v1.1.40-2
**Platforms** - Desktop(MAC), Mobile-web (Safari + Voiceover)

**Note** - Due to installation of build not able to check on Android


https://user-images.githubusercontent.com/97600950/157190811-a5857086-8788-4912-94e6-a150e4fd0a0b.mp4
